src/utilities/mergeAST.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict
+ */
+
+import { visit } from '../language/visitor';
+import {
+  type DocumentNode,
+  type FragmentDefinitionNode,
+  type FragmentSpreadNode,
+} from '../language/ast';
+
+/**
+ * Given a document AST, merge all fragment definitions into the
+ * provided operations using inline fragments.
+ */
+
+export function mergeAST(ast: DocumentNode): DocumentNode {
+  const fragmentDefinitions: { [name: string]: FragmentDefinitionNode } = {};
+  const inlinedFragments: { [name: string]: FragmentSpreadNode } = {};
+
+  visit(ast, {
+    FragmentDefinition(node) {
+      // Collect the existing FragmentDefinition nodes.
+      fragmentDefinitions[node.name.value] = node;
+    },
+  });
+
+  return visit(ast, {
+    FragmentSpread(node) {
+      // No repeats!
+      if (inlinedFragments[node.name.value]) {
```

Comment:
This is incorrect check, same fragment can be used in multiple places
```graphql
{
  person: {
    ...ProfileInfo
    friend: {
      ...ProfileInfo
    }
  }
}
```
Also please cover in tests.

Reply:
covered in unit tests now :)